tests/bluetooth/tester/src/l2cap.c

```diff
@@ -58,6 +59,17 @@ static int recv_cb(struct bt_l2cap_chan *l2cap_chan, struct net_buf *buf)
 	tester_send(BTP_SERVICE_ID_L2CAP, L2CAP_EV_DATA_RECEIVED,
 		    CONTROLLER_INDEX, recv_cb_buf, sizeof(*ev) + buf->len);
 
+	/* TODO should be list but since currently only 1 credit is
+	 * supported we can keep it simpler for now
+	 */
+	if (!chan->nbuf) {
+		/* no need for extra ref, as when returning EINPROGRESS user
```

Comment:
That the ownership of the reference `buf` depends on the return value of this callback is such an unexpected contract. It's not documented on `struct bt_l2cap_chan_ops :: recv`. Are we sure that it's like this? Can you confirm, @joerchan? I would expect either that the reference `buf` is always moved or always borrowed into this function.

By "Moved", I mean caller gave away ownership by calling and never uses this copy of the reference again not free it and must assume it has already been `net_buf_unref`ed.

By "Borrowed" I mean caller assumes it has ownership of the reference when the control returns and this function has to copy the reference (i.e. by net_buf_ref) if it wants it to live beyond this function.

Reply:
It is quite clear to me. With EINPROGRESS netbuf needs to be returned either via bt_l2cap_chan_recv_complete() or with unref (eg upon disconnection).